src/loader/mod.rs

```diff
@@ -0,0 +1,650 @@
+            None => GuestAddress(ehdr.e_entry as u64),
+        };
+
+        kernel_image
+            .seek(SeekFrom::Start(ehdr.e_phoff))
+            .map_err(|_| Error::SeekProgramHeader)?;
+        let phdrs: Vec<elf::Elf64_Phdr> = unsafe {
+            // Reading the structs is safe for a slice of POD structs.
+            struct_util::read_struct_slice(kernel_image, ehdr.e_phnum as usize)
+                .map_err(|_| Error::ReadProgramHeader)?
+        };
+
+        // Read in each section pointed to by the program headers.
+        for phdr in &phdrs {
+            if phdr.p_type != elf::PT_LOAD || phdr.p_filesz == 0 {
+                continue;
+            }
+
+            kernel_image
+                .seek(SeekFrom::Start(phdr.p_offset))
+                .map_err(|_| Error::SeekKernelStart)?;
+
+            // if the vmm does not specify where the kernel should be loaded, just
+            // load it to the physical address p_paddr for each segment.
+            let mem_offset = match kernel_start {
+                Some(start) => start
+                    .checked_add(phdr.p_paddr as u64)
+                    .ok_or(Error::InvalidProgramHeaderAddress)?,
+                None => GuestAddress(phdr.p_paddr as u64),
+            };
+
+            guest_mem
+                .read_exact_from(mem_offset, kernel_image, phdr.p_filesz as usize)
+                .map_err(|_| Error::ReadKernelImage)?;
+
+            loader_result.kernel_end = mem_offset
+                .raw_value()
+                .checked_add(phdr.p_memsz as GuestUsize)
+                .ok_or(Error::MemoryOverflow)?;
+        }
+
+        // elf image has no setup_header which is defined for bzImage
+        loader_result.setup_header = None;
+
+        Ok(loader_result)
+    }
+}
+
+#[cfg(feature = "bzimage")]
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+/// Big zImage (bzImage) kernel image support.
+pub struct BzImage;
+
+#[cfg(feature = "bzimage")]
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+impl KernelLoader for BzImage {
+    /// Loads a bzImage
+    ///
+    /// kernel is loaded into guest memory at code32_start the default load address
+    /// stored in bzImage setup header.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem` - The guest memory where the kernel image is loaded.
+    /// * `kernel_start` - The offset into 'guest_mem' at which to load the kernel.
+    /// * `kernel_image` - Input bzImage image.
+    /// * `highmem_start_address` - This is the start of the high memory, kernel should above it.
+    ///
+    /// # Returns
+    /// * KernelLoaderResult
+    fn load<F, M: GuestMemory>(
+        guest_mem: &M,
+        kernel_start: Option<GuestAddress>,
+        kernel_image: &mut F,
+        highmem_start_address: Option<GuestAddress>,
+    ) -> Result<KernelLoaderResult>
+    where
+        F: Read + Seek,
+    {
+        let mut kernel_size = kernel_image
```

Comment:
I think you're missing this step for the bzImage loading process

From https://www.kernel.org/doc/Documentation/x86/boot.txt

```In 32-bit boot protocol, the first step in loading a Linux kernel
should be to setup the boot parameters (struct boot_params,
traditionally known as "zero page"). The memory for struct boot_params
should be allocated and initialized to all zero. Then the setup header
from offset 0x01f1 of kernel image on should be loaded into struct
boot_params and examined```

The load function function prototype should probably be modified to take a `boot_params`

Reply:
I missed that this header is being returned which should be sufficient to satisfy this requirement.